File: medium/javascript/0394-decode-string.js
/**
 * @param {string} s
 * @return {string}
 */
var decodeString = function (s) {
  const stack = [];

  for (let i = 0; i < s.length; i++) {
    if (s.charAt(i) !== "]") {
      stack.push(s.charAt(i));
      continue;
    }

    let str = "";
    let currChar = stack.pop();
    while (currChar !== "[") {
      str = currChar + str;
      currChar = stack.pop();
    }

    let k = "";
    currChar = stack.pop();
    while (!Number.isNaN(Number(currChar))) {
      k = currChar + k;
      currChar = stack.pop();
    }

    stack.push(currChar);
    stack.push(str.repeat(Number(k)));
  }

  return stack.join("");
};

File: medium/go/0394-decode-string.go
package main

import (
	"strconv"
	"strings"
)

func decodeString(s string) string {
	stack := []byte{}

	for i := 0; i < len(s); i++ {
		if s[i] != ']' {
			stack = append(stack, s[i])
			continue
		}

		str := ""
		for stack[len(stack)-1] != '[' {
			str = string(stack[len(stack)-1]) + str
			stack = stack[:len(stack)-1]
		}

		stack = stack[:len(stack)-1]

		k := ""
		for len(stack) > 0 {
            _, err := strconv.Atoi(string(stack[len(stack)-1]))

			if err != nil {
				break;
			}

			k = string(stack[len(stack)-1]) + k
			stack = stack[:len(stack)-1]
		}

		if num, err := strconv.Atoi(k); err == nil {
         parsed := strings.Repeat(str, num)
		 for i := 0; i < len(parsed); i++ {
			stack = append(stack, parsed[i])
		 }
		} else {
			return err.Error();
		}
	}

	return string(stack)
}

